firmware/bb_frc_controller_pololu_32u4_robot_controller/bb_frc_controller_pololu_32u4_robot_controller.ino

```diff
@@ -1,19 +1,377 @@
+unsigned int samplingInterval = 19; // how often to run the main loop (in ms)
+
+Servo servos[BRD_MAX_SERVOS];
+byte detachedServoCount = 0;
+byte servoCount = 0;
+
+boolean isResetting = false;
+
+void outputDigitalPort(byte portValue, byte forceSend) {
+    // Pins not configured as INPUT are cleared to zeros
+    portValue = portValue & digitalPortConfigInput;
+
+    if (forceSend || previousDigitalPins != portValue) {
+        Firmata.sendDigitalPort(0, portValue);
+        previousDigitalPins = portValue;
+    }
+}
+
+byte readDigitalPort() {
+    byte portValue = 0;
+    for (byte i = 0; i < BRD_NUM_DIGITAL_PINS; i++) {
+        portValue |= ((digitalRead(digitalPinMap[i]) == HIGH ? 1 : 0) << i);
+    }
+
+    return portValue;
+}
+
+void checkDigitalInputs(void) {
+    // Since we only have 1 virtual port, just do a naive digitalRead on all of it
+    byte portValue = readDigitalPort();
+
+    outputDigitalPort(portValue, false);
+}
+
+/**
+ * Sets the pin mode to the correct state and sets the relevant bits in the
+ * bit-array that tracks Digital I/O status.
+ *
+ * We only allow setting modes on Digital IO pins here. All analog pins will
+ * remain as analog, and all servo pins remain as servos
+ */
+void setPinModeCallback(byte pin, int mode) {
+    // TODO: Maybe we don't need this function? This is a custom board with very
+    // specific layout. The only thing we need to do is to set the direction of
+    // the DIO pins.
+    if (BRD_IS_DIGITAL_PIN(pin)) {
+        if (mode == INPUT || mode == PIN_MODE_PULLUP) {
+            digitalPortConfigInput |= (1 << (pin & 7));
+        }
+        else {
+            digitalPortConfigInput &- (1 << (pin & 7));
+        }
+    }
+
+    Firmata.setPinState(pin, 0);
+
+    switch (mode) {
+        case PIN_MODE_ANALOG:
+            if (BRD_IS_ANALOG_PIN(pin)) {
+                Firmata.setPinMode(pin, PIN_MODE_ANALOG);
+            }
+            break;
+        case INPUT:
+            if (BRD_IS_DIGITAL_PIN(pin)) {
+                pinMode(digitalPinMap[pin], INPUT); // disable output driver
+                Firmata.setPinMode(pin, INPUT);
+            }
+            break;
+        case PIN_MODE_PULLUP:
+            if (BRD_IS_DIGITAL_PIN(pin)) {
+                pinMode(digitalPinMap[pin], INPUT_PULLUP);
+                Firmata.setPinMode(pin, PIN_MODE_PULLUP);
+                Firmata.setPinState(pin, 0);
+            }
+            break;
+        case OUTPUT:
+            if (BRD_IS_DIGITAL_PIN(pin)) {
+                pinMode(digitalPinMap[pin], OUTPUT);
+                Firmata.setPinMode(pin, OUTPUT);
+            }
+        default:
+            Firmata.sendString("Unsupported pin mode");
+    }
+}
+
+void setPinValueCallback(byte pin, int value) {
+    if (BRD_IS_DIGITAL_PIN(pin)) {
+        if (Firmata.getPinMode(pin) == OUTPUT) {
+            Firmata.setPinState(pin, value);
+            digitalWrite(digitalPinMap[pin], value);
+        }
+    }
```

Comment:
It would be good to have an else clause error, but that can be added in the next pull request

Reply:
Yeah I’ll add that in the next PR